# Imports

In [14]:
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [15]:
#base
import pandas as pd
import numpy as np
from scipy import stats
# visual
import matplotlib.pyplot as plt
import seaborn as sns
#path
from pathlib import Path
import sys

In [22]:
#Config
PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))
from src.config import load_config
cfg = load_config(PROJECT_ROOT / "configs" / "config.yaml")

#settings
RANDOM_STATE = cfg.data.random_state

#data_paths
TARGET_PATH = cfg.paths.raw_target_path(PROJECT_ROOT)
USERS_PATH = cfg.paths.raw_users_path(PROJECT_ROOT)

#src
sys.path.append(str(PROJECT_ROOT / cfg.paths.src_dir))

In [23]:
#src
from eda import full_eda_report, analyze_numeric, plot_distributions, plot_correlation_matrix, analyze_categorical, detect_outliers
from visual import setup_plotting

In [26]:
#Matplotlib_Settings
%matplotlib inline
setup_plotting()

# Data Loading

In [25]:
assert USERS_PATH.exists(), f'File not founded: {USERS_PATH}'
df = pd.read_csv(USERS_PATH)
df.head()

C:\Users\Максим\AppData\Local\Temp\ipykernel_3768\3232408817.py:1: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(USERS_PATH)


,session_id,client_id,visit_date,visit_time,visit_number,utm_source,utm_medium,utm_campaign,utm_adcontent,utm_keyword,device_category,device_os,device_brand,device_model,device_screen_resolution,device_browser,geo_country,geo_city
0,9055434745589932991.1637753792.1637753792,2108382700.163776,2021-11-24,14:36:32,1,ZpYIoDJMcFzVoPFsHGJL,banner,LEoPHuyFvzoNfnzGgfcd,vCIpmpaGBnIQhyYNkXqp,puhZPIYqKXeFPaUviSjo,mobile,Android,Huawei,NaN,360x720,Chrome,Russia,Zlatoust
1,905544597018549464.1636867290.1636867290,210838531.163687,2021-11-14,08:21:30,1,MvfHsxITijuriZxsqZqt,cpm,FTjNLDyTrXaWYgZymFkV,xhoenQgDQsgfEPYNPwKO,IGUCNvHlhfHpROGclCit,mobile,Android,Samsung,NaN,385x854,Samsung Internet,Russia,Moscow
2,9055446045651783499.1640648526.1640648526,2108385331.164065,2021-12-28,02:42:06,1,ZpYIoDJMcFzVoPFsHGJL,banner,LEoPHuyFvzoNfnzGgfcd,vCIpmpaGBnIQhyYNkXqp,puhZPIYqKXeFPaUviSjo,mobile,Android,Huawei,NaN,360x720,Chrome,Russia,Krasnoyarsk
3,9055447046360770272.1622255328.1622255328,2108385564.162225,2021-05-29,05:00:00,1,kjsLglQLzykiRbcDiGcD,cpc,NaN,NOBKLgtuvqYWkXQHeYWM,NaN,mobile,NaN,Xiaomi,NaN,393x786,Chrome,Russia,Moscow
4,9055447046360770272.1622255345.1622255345,2108385564.162225,2021-05-29,05:00:00,2,kjsLglQLzykiRbcDiGcD,cpc,NaN,NaN,NaN,mobile,NaN,Xiaomi,NaN,393x786,Chrome,Russia,Moscow


In [27]:
assert TARGET_PATH.exists(), f'File not founded: {TARGET_PATH}'
targets = pd.read_csv(TARGET_PATH)
targets.head()

,session_id,hit_date,hit_time,hit_number,hit_type,hit_referer,hit_page_path,event_category,event_action,event_label,event_value
0,5639623078712724064.1640254056.1640254056,2021-12-23,597864.0,30,event,NaN,sberauto.com/cars?utm_source_initial=google&ut...,quiz,quiz_show,NaN,NaN
1,7750352294969115059.1640271109.1640271109,2021-12-23,597331.0,41,event,NaN,sberauto.com/cars/fiat?city=1&city=18&rental_c...,quiz,quiz_show,NaN,NaN
2,885342191847998240.1640235807.1640235807,2021-12-23,796252.0,49,event,NaN,sberauto.com/cars/all/volkswagen/polo/e994838f...,quiz,quiz_show,NaN,NaN
3,142526202120934167.1640211014.1640211014,2021-12-23,934292.0,46,event,NaN,sberauto.com/cars?utm_source_initial=yandex&ut...,quiz,quiz_show,NaN,NaN
4,3450086108837475701.1640265078.1640265078,2021-12-23,768741.0,79,event,NaN,sberauto.com/cars/all/mercedes-benz/cla-klasse...,quiz,quiz_show,NaN,NaN


# DF EDA

In [28]:
num_cols, cat_cols = full_eda_report(df)

DATASET OVERVIEW
Shape: 1,860,042 rows x 18 columns
Memory: 1691.21 MB
Duplicates: 0 (0.00%)

--- DATA TYPES ---
  object: 17 columns
  int64: 1 columns

--- MISSING VALUES ---
                 Count  Percent
device_model   1843704    99.12
utm_keyword    1082061    58.17
device_os      1070138    57.53
device_brand    367178    19.74
utm_adcontent   335615    18.04
utm_campaign    219603    11.81
utm_source          97     0.01

--- NUMERIC COLUMNS ---
  1 columns: ['visit_number']

--- CATEGORICAL COLUMNS ---
  17 columns: ['session_id', 'client_id', 'visit_date', 'visit_time', 'utm_source', 'utm_medium', 'utm_campaign', 'utm_adcontent', 'utm_keyword', 'device_category', 'device_os', 'device_brand', 'device_model', 'device_screen_resolution', 'device_browser', 'geo_country', 'geo_city']


In [33]:
analyze_categorical(df)


Column: session_id | Unique: 1860042 | Missing: 0
  Too many categories (1860042), showing top 10:
session_id
9055434745589932991.1637753792.1637753792    1
905544597018549464.1636867290.1636867290     1
9055446045651783499.1640648526.1640648526    1
9055447046360770272.1622255328.1622255328    1
9055447046360770272.1622255345.1622255345    1
9055447192389856083.1622453074.1622453074    1
9055455318486370642.1640843788.1640843788    1
9055461992850812764.1626107740.1626107740    1
9055462349345527315.1638536723.1638536723    1
9055466554104774132.1624800757.1624800757    1

Column: client_id | Unique: 1391717 | Missing: 0
  Too many categories (1391717), showing top 10:
client_id
1750498477.162945    462
130918396.162801     334
1266574870.162314    309
736859557.161956     306
1894772482.162186    276
1065181919.163161    271
1136708344.163342    257
1761941933.162184    254
260799907.1631       252
943676978.162189     248

Column: visit_date | Unique: 226 | Missing: 0
  Too many ca

#ВЫВОДЫ:

## Типы данных:
Все фичи являются типо object, однако имеются такие колонки как visit_date, visit_time которые могут/должны быть преобразованы в date, time соответственно. Device screen resolution можно распарсить в ширину и высоту а также площадь что бы 

## Missing Valuse:
- Device_model: 99.12% пропусков + закодирована => удаляем фичу
- utm_keyword: ключевое слово 58.17% + закодирована => восстановить невозможно, но есть доминирующее ключевое слово, можно сделать фичу является ли ключевое слово доминантным или нет(пропуски = нет)
- device_os: 57.53% пропусков, есть 2 варианта заполнения т.к. есть 3 доминирующих операционных системы можно заполнить с сохранением пропорции, но более чистый вариант будет либо удалить, либо заполнить (not set)
- device_brand: 19,74% пропусков, нет ярко выраженного доминантного производителя, заполню (not set)
- utm_adcontent: 18.04% пропусков + закодирована, есть доминантный представитель, можно заполнить им, либо ввести новую категорию (not set) + 286 уникальных категорий из которых большую половину занимает 1 категорию, можно добавить фичу принадлежит ли данной фиче
- utm_campaign: 11.81% + закодирована: нет доминирующей категории => либо распределить в соответствии с пропорциями первых 4 т.к. они в сумме дают высокий процент по фиче либо удалить эти объекты
- utm_source: 0.01% + закодирована: либо удалить объекты либо попробовать восстановить из других закодированных фич где совпадения такие же как и в пропусках(например utm_medium + utm_campaign)

## Наличие доминантных категорий
- utm_adcontent
- utm_keyword
- device_os(android+ iOS+ Windows)
- 